### DISCUSSION TITLE
🔄 PIVOT > Make it possible to create VIEW as select * from {PIVOTED statement}

### DISCUSSION THREAD

**Discussion by adriens** [duckdb/duckdb#8947] [Ideas]
## :grey_question: About

On a project I needed to get a PIVOT output as a table or `csv` file.

:point_up: Actually, we cannot : 

```sql
create or replace view pivot_vw_CONSUMPTION_SUBSECTOR as select * from(
    pivot vw_CONSUMPTION_SUBSECTOR
    on CONSUMPTION_SUBSECTOR
    using median(median_price)
    order by year_month asc);
```

as it leads to:

<pre>Error: Parser Error: PIVOT statements with pivot elements extracted from the data cannot be used in views.
In order to use PIVOT in a view the PIVOT values must be manually specified, e.g.:
PIVOT ... ON CONSUMPTION_SUBSECTOR IN (val1, val2, ...)
D </pre>

## :nerd_face: Workaround

To achieve my project I had to, from Python:

1. Put the `PIVOT` output into a `df`
2. Dump the `df` into a `csv` file
3. Do anything with the `csv` (including loading it into `duckdb` :sweat_smile: )

## :dart: Feature request

- Make it possible to create a VIEW on top of pivoted statements

**Comment by l1t1** [comment]
remove `select * ` works
```
D SET pivot_limit=110000;
Run Time (s): real 0.002 user 0.000247 sys 0.000265
D create table t100000x100 as
> pivot
> (
> select i::int i,(i%100000)::int j,(i//100000)::int16 k from range(100000*100)t(i)
> )
> on j using first(i)group by k;
```

**Comment by l1t1** [comment]
add `()` also works
```

create table t2x5 as select "0","1" from(
pivot
(
select i,i%3 j,i//3 k from range(3*5)t(i)
)
on j using sum(i)group by k);


create table t3x5 as select * from(
pivot
(
select i,i%3 j,i//3 k from range(3*5)t(i)
)
on j using sum(i)group by k);
```

**Reply by adriens** [reply]
I'mgiving it a try within a few hours :rocket:

**Reply by adriens** [reply]
:point_right: You're right for the table..., but the `VIEW` won't work

**Comment by l1t1** [comment]
view also works, the rule is strict. 
```sql

D CREATE TABLE weather (
>     city    VARCHAR,
>     temp_lo INTEGER, -- minimum temperature on a day
>     temp_hi INTEGER, -- maximum temperature on a day
>     prcp    REAL,
>     date    DATE
> );
D
D CREATE TABLE cities (
>     city VARCHAR,
>     lat  DECIMAL,
>     lon  DECIMAL
> );
D
D INSERT INTO weather VALUES ('San Francisco', 46, 50, 0.25, '1994-11-27');
D INSERT INTO weather VALUES ('San Francisco', 43, 57, 0.0 , '1994-11-29');
D INSERT INTO weather VALUES ('Hayward'      , 54, 37, null, '1994-11-29');
D
D INSERT INTO cities VALUES ('San Francisco', -194.0, 53.0);
D pivot weather on city using sum(prcp);
┌─────────┬─────────┬────────────┬─────────┬───────────────┐
│ temp_lo │ temp_hi │    date    │ Hayward │ San Francisco │
│  int32  │  int32  │    date    │ double  │    double     │
├─────────┼─────────┼────────────┼─────────┼───────────────┤
│      43 │      57 │ 1994-11-29 │         │           0.0 │
│      46 │      50 │ 1994-11-27 │         │          0.25 │
│      54 │      37 │ 1994-11-29 │         │               │
└─────────┴─────────┴────────────┴─────────┴───────────────┘
D create view v as pivot weather on city using sum(prcp);
Error: Parser Error: PIVOT statements with pivot elements extracted from the data cannot be used in views.
In order to use PIVOT in a view the PIVOT values must be manually specified, e.g.:
PIVOT ... ON city IN (val1, val2, ...)
D create view v as pivot weather on city in ('San Francisco','Hayward')using sum(prcp);
D from v;
┌─────────┬─────────┬────────────┬───────────────┬─────────┐
│ temp_lo │ temp_hi │    date    │ San Francisco │ Hayward │
│  int32  │  int32  │    date    │    double     │ double  │
├─────────┼─────────┼────────────┼───────────────┼─────────┤
│      43 │      57 │ 1994-11-29 │           0.0 │         │
│      46 │      50 │ 1994-11-27 │          0.25 │         │
│      54 │      37 │ 1994-11-29 │               │         │
└─────────┴─────────┴────────────┴───────────────┴─────────┘
```

**Comment by l1t1** [comment]
you can get the values by
```
D select list(distinct ''''||city||'''')from weather;
┌─────────────────────────────────────────┐
│ list(DISTINCT (('''' || city) || '''')) │
│                varchar[]                │
├─────────────────────────────────────────┤
│ ['San Francisco', 'Hayward']            │
└─────────────────────────────────────────┘
```

**Comment by adriens** [comment]
Hmm, that would be very easy for end user to experiment the same Dev. Experience while playing with tables than with views

<pre>D <font color="#8AE234"><b>select </b></font>* <font color="#8AE234"><b>from</b></font>(
&gt;         <font color="#8AE234"><b>pivot </b></font>vw_CONSUMPTION_SUBSECTOR
&gt;         <font color="#8AE234"><b>on </b></font>CONSUMPTION_SUBSECTOR
&gt;         <font color="#8AE234"><b>using </b></font>median(median_price)
&gt;         <font color="#8AE234"><b>order by </b></font>year_month <font color="#8AE234"><b>asc</b></font>
&gt;     );
┌────────────┬──────────────────────┬──────────────────────┬───────────────────┬──────────────────────┬───┬──────────────────────┬──────────────────┬──────────────────────┬──────────────────────┬─────────────────┬────────┐
│ year_month │ apéritif, petit fo…  │ assaisonnement, co…  │ beurre, margarine │ biscotte, tartine …  │ … │ sucre, farine, aid…  │ tapas, antipasti │ traiteur, rôtisserie │ viande, volaille s…  │ volaille, lapin │ yaourt │
│  varchar   │        double        │        double        │      double       │        double        │   │        double        │      double      │        double        │        double        │     double      │ double │
├────────────┼──────────────────────┼──────────────────────┼───────────────────┼──────────────────────┼───┼──────────────────────┼──────────────────┼──────────────────────┼──────────────────────┼─────────────────┼────────┤
│ 2014-01    │                      │                310.0 │             294.0 │               348.25 │ … │                310.0 │                  │               2620.0 │                450.0 │                 │  363.0 │
│ 2014-02    │                      │                315.0 │             294.0 │               275.75 │ … │                315.0 │                  │               2620.0 │                450.0 │                 │  351.0 │
│ 2014-03    │                      │                315.0 │             295.0 │                306.5 │ … │                320.0 │                  │               2620.0 │                440.0 │                 │  330.0 │
│ 2014-04    │                970.0 │                312.5 │             300.0 │                395.0 │ … │               320.75 │                  │              1713.75 │               450.25 │                 │  377.5 │
│ 2014-05    │                      │                310.0 │             278.0 │                395.0 │ … │                325.0 │                  │               1632.5 │                470.0 │                 │  395.0 │
│ 2014-06    │                970.0 │                315.0 │             305.0 │                395.0 │ … │                335.0 │                  │               1550.0 │                510.0 │                 │  395.0 │
│ 2014-07    │                      │                310.0 │             311.0 │                365.0 │ … │                335.0 │                  │               1185.0 │                460.0 │                 │  384.5 │
│ 2014-08    │                      │                310.0 │             281.5 │                390.0 │ … │                325.0 │                  │               1180.0 │                510.0 │                 │  380.0 │
│ 2014-09    │                885.0 │                315.0 │             350.0 │               292.75 │ … │                350.0 │                  │               1150.0 │               506.25 │                 │  369.0 │
│ 2014-10    │                      │                295.0 │             348.5 │                375.0 │ … │                330.0 │                  │                995.0 │                475.0 │                 │  366.5 │
│ 2014-11    │                      │                310.0 │            313.75 │                390.0 │ … │                330.0 │                  │                995.0 │                471.5 │                 │  390.0 │
│ 2014-12    │                      │                322.5 │             344.0 │                390.0 │ … │                347.5 │                  │               1190.0 │                525.0 │                 │  395.0 │
│ 2023-01    │               1162.5 │                352.5 │             465.0 │                320.0 │ … │                377.5 │            350.0 │              1141.25 │                885.0 │          2830.0 │  395.0 │
│ 2023-02    │               1025.0 │                350.0 │             470.0 │                320.0 │ … │                375.0 │            379.5 │              1108.75 │                867.0 │          2505.0 │  395.0 │
│ 2023-03    │               1195.0 │                375.0 │             473.5 │                325.0 │ … │                389.5 │            379.5 │              1058.75 │                795.0 │          2530.0 │  395.0 │
│ 2023-04    │               1245.0 │                375.0 │             482.5 │                325.0 │ … │                375.0 │            350.0 │              1012.75 │                795.0 │          2830.0 │  395.0 │
│ 2023-05    │               1025.0 │                369.5 │             493.5 │                340.0 │ … │               391.25 │            350.0 │               1059.0 │                797.5 │          2762.5 │  395.0 │
│ 2023-06    │               1025.0 │               369.25 │             495.0 │                350.5 │ … │                385.5 │           346.25 │               1120.0 │                735.0 │          2762.5 │ 398.75 │
│ 2023-07    │                850.0 │                390.0 │             485.0 │                352.5 │ … │                390.0 │            350.0 │               1090.0 │                682.5 │          2675.0 │  402.0 │
│ 2023-08    │                850.0 │                355.0 │             485.0 │                370.0 │ … │                385.0 │            350.0 │                930.0 │                795.0 │          2830.0 │  420.0 │
├────────────┴──────────────────────┴──────────────────────┴───────────────────┴──────────────────────┴───┴──────────────────────┴──────────────────┴──────────────────────┴──────────────────────┴─────────────────┴────────┤
│ 20 rows                                                                                                                                                                                              53 columns (11 shown) │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
D <font color="#8AE234"><b>create or replace view </b></font>pivot_vw_CONSUMPTION_SUBSECTOR <font color="#8AE234"><b>as</b></font>
&gt;     <font color="#8AE234"><b>select </b></font>year_month, yaourt <font color="#8AE234"><b>from</b></font>(
&gt;         <font color="#8AE234"><b>pivot </b></font>vw_CONSUMPTION_SUBSECTOR
&gt;         <font color="#8AE234"><b>on </b></font>CONSUMPTION_SUBSECTOR
&gt;         <font color="#8AE234"><b>using </b></font>median(median_price)
&gt;         <font color="#8AE234"><b>order by </b></font>year_month <font color="#8AE234"><b>asc</b></font>
&gt;     );
Error: Parser Error: PIVOT statements with pivot elements extracted from the data cannot be used in views.
In order to use PIVOT in a view the PIVOT values must be manually specified, e.g.:
PIVOT ... ON CONSUMPTION_SUBSECTOR IN (val1, val2, ...)
D 

</pre>
